File: php/test/ProtectedBranchPullRequestReviewEntityTest.php
<?php
declare(strict_types=1);

// ProtectedBranchPullRequestReview entity test

require_once __DIR__ . '/../github_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class ProtectedBranchPullRequestReviewEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = GithubSDK::test(null, null);
        $ent = $testsdk->ProtectedBranchPullRequestReview(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = protected_branch_pull_request_review_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["update", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "protected_branch_pull_request_review." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $protected_branch_pull_request_review_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.protected_branch_pull_request_review")));
        $protected_branch_pull_request_review_ref01_data = null;
        if (count($protected_branch_pull_request_review_ref01_data_raw) > 0) {
            $protected_branch_pull_request_review_ref01_data = Helpers::to_map($protected_branch_pull_request_review_ref01_data_raw[0][1]);
        }

        // UPDATE
        $protected_branch_pull_request_review_ref01_ent = $client->ProtectedBranchPullRequestReview(null);
        $protected_branch_pull_request_review_ref01_data_up0_up = [
            "owner" => $setup["idmap"]["owner"],
            "repo" => $setup["idmap"]["repo"],
        ];

        $protected_branch_pull_request_review_ref01_markdef_up0_name = "url";
        $protected_branch_pull_request_review_ref01_markdef_up0_value = "Mark01-protected_branch_pull_request_review_ref01_" . $setup["now"];
        $protected_branch_pull_request_review_ref01_data_up0_up[$protected_branch_pull_request_review_ref01_markdef_up0_name] = $protected_branch_pull_request_review_ref01_markdef_up0_value;

        $protected_branch_pull_request_review_ref01_resdata_up0_result = $protected_branch_pull_request_review_ref01_ent->update($protected_branch_pull_request_review_ref01_data_up0_up, null);
        $protected_branch_pull_request_review_ref01_resdata_up0 = Helpers::to_map(is_object($protected_branch_pull_request_review_ref01_resdata_up0_result) && method_exists($protected_branch_pull_request_review_ref01_resdata_up0_result, 'data_get') ? $protected_branch_pull_request_review_ref01_resdata_up0_result->data_get() : $protected_branch_pull_request_review_ref01_resdata_up0_result);
        $this->assertNotNull($protected_branch_pull_request_review_ref01_resdata_up0);
        $this->assertEquals($protected_branch_pull_request_review_ref01_resdata_up0[$protected_branch_pull_request_review_ref01_markdef_up0_name], $protected_branch_pull_request_review_ref01_markdef_up0_value);

        // LOAD
        $protected_branch_pull_request_review_ref01_match_dt0 = [];
        $protected_branch_pull_request_review_ref01_data_dt0_loaded = $protected_branch_pull_request_review_ref01_ent->load($protected_branch_pull_request_review_ref01_match_dt0, null);
        $this->assertNotNull($protected_branch_pull_request_review_ref01_data_dt0_loaded);

    }
}

function protected_branch_pull_request_review_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/protected_branch_pull_request_review/ProtectedBranchPullRequestReviewTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = GithubSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["protected_branch_pull_request_review01", "protected_branch_pull_request_review02", "protected_branch_pull_request_review03", "repo01", "repo02", "repo03", "branch01", "branch02", "branch03", "owner01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID" => $idmap,
        "GITHUB_TEST_LIVE" => "FALSE",
        "GITHUB_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["owner"])) {
        $idmap_resolved["owner"] = $idmap_resolved["owner01"];
    }
    if (!isset($idmap_resolved["repo"])) {
        $idmap_resolved["repo"] = $idmap_resolved["repo01"];
    }

    if ($env["GITHUB_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        // "?? []" because merge legitimately answers with a stdClass when every
        // contributing entry is an EMPTY map - an SDK with no apikey and no
        // server variables generates an empty middle entry, so that is the
        // common case, not the edge one. to_map returns null for a non-array by
        // design, and the constructor takes a non-nullable array, so without the
        // fallback every such SDK died on "must be of type array, null given"
        // the moment live mode was switched on. Offline mode never reaches this
        // branch, which is why the offline suite stayed green.
        $client = new GithubSDK(Helpers::to_map($merged_opts) ?? []);
    }

    $live = $env["GITHUB_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["GITHUB_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/protected_branch_pull_request_review_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestProtectedBranchPullRequestReviewEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ProtectedBranchPullRequestReview(nil)
		if ent == nil {
			t.Fatal("expected non-nil ProtectedBranchPullRequestReviewEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := protected_branch_pull_request_reviewBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "protected_branch_pull_request_review." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		protectedBranchPullRequestReviewRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath(setup.data, "existing.protected_branch_pull_request_review")))
		var protectedBranchPullRequestReviewRef01Data map[string]any
		if len(protectedBranchPullRequestReviewRef01DataRaw) > 0 {
			protectedBranchPullRequestReviewRef01Data = core.ToMapAny(protectedBranchPullRequestReviewRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = protectedBranchPullRequestReviewRef01Data

		// UPDATE
		protectedBranchPullRequestReviewRef01Ent := client.ProtectedBranchPullRequestReview(nil)
		protectedBranchPullRequestReviewRef01DataUp0Up := map[string]any{
			"owner": setup.idmap["owner"],
			"repo": setup.idmap["repo"],
		}

		protectedBranchPullRequestReviewRef01MarkdefUp0Name := "url"
		protectedBranchPullRequestReviewRef01MarkdefUp0Value := fmt.Sprintf("Mark01-protected_branch_pull_request_review_ref01_%d", setup.now)
		protectedBranchPullRequestReviewRef01DataUp0Up[protectedBranchPullRequestReviewRef01MarkdefUp0Name] = protectedBranchPullRequestReviewRef01MarkdefUp0Value

		protectedBranchPullRequestReviewRef01ResdataUp0Result, err := protectedBranchPullRequestReviewRef01Ent.Update(protectedBranchPullRequestReviewRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		protectedBranchPullRequestReviewRef01ResdataUp0 := core.ToMapAny(entityData(protectedBranchPullRequestReviewRef01ResdataUp0Result))
		if protectedBranchPullRequestReviewRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if protectedBranchPullRequestReviewRef01ResdataUp0[protectedBranchPullRequestReviewRef01MarkdefUp0Name] != protectedBranchPullRequestReviewRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", protectedBranchPullRequestReviewRef01MarkdefUp0Name, protectedBranchPullRequestReviewRef01ResdataUp0[protectedBranchPullRequestReviewRef01MarkdefUp0Name])
		}

		// LOAD
		protectedBranchPullRequestReviewRef01MatchDt0 := map[string]any{}
		protectedBranchPullRequestReviewRef01DataDt0Loaded, err := protectedBranchPullRequestReviewRef01Ent.Load(protectedBranchPullRequestReviewRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if protectedBranchPullRequestReviewRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func protected_branch_pull_request_reviewBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "protected_branch_pull_request_review", "ProtectedBranchPullRequestReviewTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read protected_branch_pull_request_review test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse protected_branch_pull_request_review test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"protected_branch_pull_request_review01", "protected_branch_pull_request_review02", "protected_branch_pull_request_review03", "repo01", "repo02", "repo03", "branch01", "branch02", "branch03", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_PROTECTED_BRANCH_PULL_REQUEST_REVIEW_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}
	// Add repo alias for update test.
	if idmapResolved["repo"] == nil {
		idmapResolved["repo"] = idmapResolved["repo01"]
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
